sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadViaSDF.java

```diff
@@ -0,0 +1,742 @@
+    return withExtractOutputTimestampFn(ExtractOutputTimestampFns.useCreateTime());
+  }
+
+  // If a transactional producer is used and it's desired to only read records from committed
+  // transaction, it's recommended to set read_committed. Otherwise, read_uncommitted is the default
+  // value.
+  public ReadViaSDF<K, V> withReadCommitted() {
+    return withConsumerConfigUpdates(ImmutableMap.of("isolation.level", "read_committed"));
+  }
+
+  public ReadViaSDF<K, V> commitOffsets() {
+    return toBuilder().setCommitOffsetEnabled(true).build();
+  }
+
+  public ReadViaSDF<K, V> withOffsetConsumerConfigOverrides(
+      Map<String, Object> offsetConsumerConfig) {
+    return toBuilder().setOffsetConsumerConfig(offsetConsumerConfig).build();
+  }
+
+  public ReadViaSDF<K, V> withConsumerConfigOverrides(Map<String, Object> consumerConfig) {
+    return toBuilder().setConsumerConfig(consumerConfig).build();
+  }
+
+  @Override
+  public PCollection<KafkaRecord<K, V>> expand(PCollection<Row> input) {
+    checkArgument(
+        ExperimentalOptions.hasExperiment(input.getPipeline().getOptions(), "beam_fn_api"),
+        "The ReadFromKafkaViaSDF can only used when beam_fn_api is enabled.");
+
+    checkArgument(getKeyDeserializerProvider() != null, "withKeyDeserializer() is required");
+    checkArgument(getValueDeserializerProvider() != null, "withValueDeserializer() is required");
+
+    ConsumerSpEL consumerSpEL = new ConsumerSpEL();
+    if (!consumerSpEL.hasOffsetsForTimes()) {
+      LOG.warn(
+          "Kafka client version {} is too old. Versions before 0.10.1.0 are deprecated and "
+              + "may not be supported in next release of Apache Beam. "
+              + "Please upgrade your Kafka client version.",
+          AppInfoParser.getVersion());
+    }
+
+    if (isCommitOffsetEnabled()) {
+      if (configuredKafkaCommit()) {
+        LOG.info(
+            "Either read_committed or auto_commit is set together with commitOffsetEnabled but you "
+                + "only need one of them. The commitOffsetEnabled is going to be ignored");
+      }
+    }
+
+    if (getConsumerConfig().get(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG) == null) {
+      LOG.warn(
+          "The bootstrapServers is not set. Then it must be populated through KafkaSourceDescription during runtime. Otherwise, the pipeline will fail.");
+    }
+
+    CoderRegistry coderRegistry = input.getPipeline().getCoderRegistry();
+    Coder<K> keyCoder = getKeyCoder(coderRegistry);
+    Coder<V> valueCoder = getValueCoder(coderRegistry);
+    Coder<KafkaRecord<K, V>> outputCoder = KafkaRecordCoder.of(keyCoder, valueCoder);
+    PCollection<KafkaRecord<K, V>> output =
+        input.apply(ParDo.of(new ReadFromKafkaDoFn())).setCoder(outputCoder);
+    if (isCommitOffsetEnabled() && !configuredKafkaCommit()) {
+      // TODO(BEAM-10123): Add CommitOffsetTransform to expansion.
+      LOG.warn("Offset committed is not supported yet. Ignore the value.");
```

Comment:
Shouldn't this be a construction time error instead of a runtime warning?

Reply:
Yes, it should be an error.